### ISSUE TITLE
Use PERF_RECORD_COMM

### ISSUE THREAD

**Issue by tilsche** [tud-zih-energy/lo2s#78]
`PERF_RECORD_COMM` provides information about name changes of processes **and threads**. This should allow better naming information. Requires Linux 3.16, so probably only optional for now.

**Comment by cvonelm** [comment]
@phijor have you already started working on this? #9 requires a different approach to process comms, as they are needed by the sample writer, but are currently set in the ProcessMonitor, which is apparently not present in system mode. Otherwise I would like to take over this, and integrate it directly with the sampling in system mode changes

**Comment by phijor** [comment]
I have indeed, but I'm not sure whether I found a suitable place for the code. It's not all that difficult to get working in principle; here's a minimal diff:

```
diff --git a/include/lo2s/perf/event_reader.hpp b/include/lo2s/perf/event_reader.hpp
index f8c3d94..3cec360 100644
--- a/include/lo2s/perf/event_reader.hpp
+++ b/include/lo2s/perf/event_reader.hpp
@@ -32,7 +32,8 @@
 #include <cstdint>
 #include <cstring>
 
-extern "C" {
+extern "C"
+{
 #include <linux/perf_event.h>
 #include <sys/mman.h>
 }
@@ -105,6 +106,15 @@ public:
         // struct sample_id sample_id;
     };
 
+    struct RecordCommType
+    {
+        struct perf_event_header header;
+        uint32_t pid;
+        uint32_t tid;
+        char comm[1]; // ISO C++ forbits zero-size array
+        // struct sample_id sample_id;
+    };
+
     ~EventReader()
     {
         if (lost_samples > 0)
@@ -215,6 +225,10 @@ public:
                     stop = crtp_this->handle((const ActualSampleType*)event_header_p);
                     break;
                 }
+                case PERF_RECORD_COMM:
+                    // Log::debug() << "PERF_RECORD_COMM";
+                    stop = crtp_this->handle((const RecordCommType*)event_header_p);
+                    break;
                 default:
                     stop = crtp_this->handle((const RecordUnknownType*)event_header_p);
                 }
@@ -307,5 +321,5 @@ private:
     void* base;
     char event_copy[PERF_SAMPLE_MAX_SIZE] __attribute__((aligned(8)));
 };
-}
-}
+} // namespace perf
+} // namespace lo2s
diff --git a/include/lo2s/perf/sample/reader.hpp b/include/lo2s/perf/sample/reader.hpp
index 8c81973..343db21 100644
--- a/include/lo2s/perf/sample/reader.hpp
+++ b/include/lo2s/perf/sample/reader.hpp
@@ -85,6 +85,9 @@ protected:
         perf_attr.clockid = config().clockid;
 #endif
 
+        perf_attr.comm = 1;
+        perf_attr.comm_exec = 1;
+
         // We need this to get all mmap_events
         if (enable_on_exec)
         {
diff --git a/include/lo2s/perf/sample/writer.hpp b/include/lo2s/perf/sample/writer.hpp
index 58f7790..37aa078 100644
--- a/include/lo2s/perf/sample/writer.hpp
+++ b/include/lo2s/perf/sample/writer.hpp
@@ -62,6 +62,7 @@ public:
     using Reader<Writer>::handle;
     bool handle(const Reader::RecordSampleType* sample);
     bool handle(const Reader::RecordMmapType* mmap_event);
+    bool handle(const Reader::RecordCommType* comm);
 
     otf2::writer::local& otf2_writer()
     {
diff --git a/lib/nitro b/lib/nitro
--- a/lib/nitro
+++ b/lib/nitro
@@ -1 +1 @@
-Subproject commit 742fd2e70e00cf8f420ef879eea09d44372d6104
+Subproject commit 742fd2e70e00cf8f420ef879eea09d44372d6104-dirty
diff --git a/lib/otf2xx b/lib/otf2xx
--- a/lib/otf2xx
+++ b/lib/otf2xx
@@ -1 +1 @@
-Subproject commit 3aabf10d7671437d5fcc521ba3eea49c566362bc
+Subproject commit 3aabf10d7671437d5fcc521ba3eea49c566362bc-dirty
diff --git a/src/perf/sample/writer.cpp b/src/perf/sample/writer.cpp
index 4af80df..64aabab 100644
--- a/src/perf/sample/writer.cpp
+++ b/src/perf/sample/writer.cpp
@@ -188,6 +188,18 @@ bool Writer::handle(const Reader::RecordMmapType* mmap_event)
     return false;
 }
 
+bool Writer::handle(const Reader::RecordCommType* comm)
+{
+    std::string new_name{ static_cast<const char*>(comm->comm) };
+
+    Log::debug() << "PERF_RECORD_COMM: comm = " << new_name << ", pid = " << comm->pid
+                 << ", tid = " << comm->pid;
+
+    trace_.process_update_executable(comm->pid, new_name);
+
+    return false;
+}
+
 void Writer::end()
 {
     // time::now() can sometimes can be in the past :-(
```

But I wouldn't mind if you took over.

**Comment by phijor** [comment]
Btw, a useful script for testing `comm` changes:
```py
#!/usr/bin/env python3

import os

def get_comm():
    with open('/proc/self/comm', 'r') as f:
        return f.read().strip()

def set_comm(comm):
    with open('/proc/self/comm', 'w') as f:
        f.write(comm)

print("on launch: " + get_comm())

for i in range(5):
    set_comm(f'[test {i}]')
    print("i = {}: {}".format(i, get_comm()))

print("exiting...")

# test for comm changes generated by exec
os.execvp('sleep', ('sleep', '2'))
```

**Comment by cvonelm** [comment]
Okay, this is close to what i would've done it, so no need for big changes.

**Comment by tilsche** [comment]
For threads, please also look at `/proc/TID/comm` similarly to how we do it for processes.

**Comment by bmario** [comment]
Fixed in ba760c2ea0f816a05fd5fd5287c382359d036e0e
